agent/engine/docker_task_engine.go

```diff
@@ -1486,6 +1582,53 @@ func (engine *DockerTaskEngine) pullAndUpdateContainerReference(task *apitask.Ta
 	return metadata
 }
 
+// Sets image registry auth credentials for the container.
+// Returns a cleanup function that may be called to clear the credentials from the container
```

Comment:
Could you help me understand what's the benefit/need for unsetting the credentials?

Reply:
Also, the credentials resources are currently mapped as resource dependencies for container like [here](https://github.com/amogh09/amazon-ecs-agent/blob/master/agent/api/task/task.go#L861-L863). I'm wondering if it's possible to do the same? It's easier and cleaner to manage all resource dependencies in one place.

Reply:
This follows existing logic from image pulls. 
https://github.com/aws/amazon-ecs-agent/blob/b702281f0f6aa2fba5614050769aa337d1456853/agent/engine/docker_task_engine.go#L1419

I think the idea is to improve Agent's security posture by clearing auth credentials from the container as soon as they are not needed (I understand that the credentials might still be available in other components).

Reply:
The `setCredentials` method gets the credentials from the relevant resource dependency of the task if the container requires credentials from ASM. There is no change to how credentials are modeled and managed.

Reply:
> I think the idea is to improve Agent's security posture by clearing auth credentials from the container as soon as they are not needed (I understand that the credentials might still be available in other components).

Okay. I am still not fully convinced why it is necessary to intentionally add this finer control granularity, when task execution role permissions are supposedly granted to all containers in the task equally. But this is non-blocking for this PR.

Reply:
> The setCredentials method gets the credentials from the relevant resource dependency of the task if the container requires credentials from ASM. There is no change to how credentials are modeled and managed.

If I'm reading correctly, the `SetCredentials` assumes that the credentials have already been retrieved remotely (from ASM), and is simply copying the creds from one internal struct to another. The actual retrieval process happens as this resource [dependency](https://github.com/amogh09/amazon-ecs-agent/blob/master/agent/api/task/task.go#L861-L863) gets fulfilled. The ASM credentials resource is currently mapped as a dependency for `ContainerPulled`, should we change it to `ManifestPulled` (or whatever the new container state is named) to ensure that the credentials have been retrieved before attempting to copy them?

Reply:
> If I'm reading correctly, the SetCredentials assumes that the credentials have already been retrieved remotely (from ASM), and is simply copying the creds from one internal struct to another.

Yes that's right. 

> The ASM credentials resource is currently mapped as a dependency for ContainerPulled, should we change it to ManifestPulled (or whatever the new container state is named) to ensure that the credentials have been retrieved before attempting to copy them?

Yes that's right. That change has already been merged. 
https://github.com/aws/amazon-ecs-agent/blob/b702281f0f6aa2fba5614050769aa337d1456853/agent/api/task/task.go#L1052-L1059